conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java

```diff
@@ -0,0 +1,99 @@
+import com.palantir.conjure.java.undertow.lib.AsyncRequestProcessing;
+import com.palantir.conjure.java.undertow.lib.ReturnValueWriter;
+import com.palantir.conjure.java.undertow.lib.Serializer;
+import com.palantir.logsafe.Preconditions;
+import io.undertow.server.ExchangeCompletionListener;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.AttachmentKey;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.xnio.XnioExecutor;
+
+final class ConjureAsyncRequestProcessing implements AsyncRequestProcessing {
+
+    private static final Executor DIRECT_EXECUTOR = MoreExecutors.directExecutor();
+    private static final AttachmentKey<ListenableFuture<?>> FUTURE = AttachmentKey.create(ListenableFuture.class);
+    // If the request is ended before the future has completed, cancel the future to signal that work
+    // should be stopped. This occurs when clients cancel requests or connections are closed.
+    private static final ExchangeCompletionListener COMPLETION_LISTENER = (exchange, nextListener) -> {
+        ListenableFuture<?> future = exchange.getAttachment(FUTURE);
+        if (future != null && !future.isDone()) {
+            future.cancel(false);
+        }
+        nextListener.proceed();
+    };
+
+    private final Serializer<SerializableError> exceptionSerializer = ConjureExceptions.serializer();
+
+    private final Duration timeout;
+
+    ConjureAsyncRequestProcessing(Duration timeout) {
+        this.timeout = Preconditions.checkNotNull(timeout, "Timeout is required");
+    }
+
+    @Override
+    public <T> void register(
+            ListenableFuture<T> future,
+            ReturnValueWriter<T> returnValueWriter,
+            HttpServerExchange exchange) throws IOException {
+        Preconditions.checkNotNull(future, "future");
+        Preconditions.checkNotNull(returnValueWriter, "returnValueWriter");
+        Preconditions.checkNotNull(exchange, "exchange");
+
+        // TODO(ckozak): potential optimization: avoid both dispatching and scheduling a timeout if the future is
+        // already complete.
+
+        // Attach data to the exchange in order to reuse a stateless listener
+        exchange.putAttachment(FUTURE, future);
+        // Cancel the future if the exchange is completed before the future is done
+        exchange.addExchangeCompleteListener(COMPLETION_LISTENER);
+
+        XnioExecutor.Key timeoutKey = exchange.getIoThread()
+                .executeAfter(() -> future.cancel(false), timeout.toMillis(), TimeUnit.MILLISECONDS);
+        future.addListener(timeoutKey::remove, DIRECT_EXECUTOR);
+        // Dispatch the registration task. We're executing inside of root handler, so the dispatched task will
+        // execute on this thread after the current handler completes execution. This allows us to avoid racing
```

Comment:
I *think* I get it, but I'm not sure; can you try to reword this explanation?
(is the problem that we need to call exchange.dispatch before the root handler terminates so that the connection isn't closed?)

Reply:
That is correct, I've updated the comment.  Does it make more sense this way?

```java
// Dispatch the registration task, this accomplishes two things:
// 1. Puts the exchange into a 'dispatched' state, otherwise the request will be terminated when
//    the endpoint HttpHandler returns. See Connectors.executeRootHandler for more information.
// 2. Avoids racing completion with the current handler chain. The dispatched task will be executed
//    after this Endpoints HttpHandler returns. Otherwise an unlucky future could race
//    HttpServerExchange.isInCall transitioning from true -> false, causing hte dispatch task not
//    to execute.
```